prboom2/src/s_sound.c

```diff
@@ -233,6 +233,14 @@ void S_Start(void)
   memset(&musinfo, 0, sizeof(musinfo));
   musinfo.items[0] = -1;
 
+  // prevent music number under- and overflows
+  if (mnum <= mus_None || (gamemode == commercial && mnum <  mus_runnin) ||
+      mnum >= NUMMUSIC || (gamemode != commercial && mnum >= mus_runnin))
+  {
+    S_StopMusic();
```

Comment:
Instead of no music, should we have a default? Like taking the first map's midi for instance.

Reply:
In Crispy, I have some code to wrap the music number around back into the valid range if it either under- or overflows. But that's not canonical. Crashing would be canonical, but not desired, and silence somehow comes closest. 
But, honestly, I don't really like it either.

Reply:
Wait a minute. It seems like it's entirely impossible to reach a music overflow in Doom 1 (at least as long as we don't explicitly support two-digit maps - or only support them "by accident" like e.g. E1M10), because for episodes >= 4 all music will get mapped into episode 4's `spmus[]` music number array. So, we are left with Doom 2's music and for this I could wrap music numbers around back into the valid range - which doesn't make a difference for any map supported so far and prevents crashes for all the maps beyond that range.

Reply:
Something like this?
```patch
--- a/prboom2/src/s_sound.c
+++ b/prboom2/src/s_sound.c
@@ -181,6 +181,15 @@ void S_Stop(void)
 // Kills playing sounds at start of level,
 //  determines music if any, changes music.
 //
+
+static inline int WRAP(int i, int w)
+{
+  while (i < 0)
+    i += w;
+
+  return i % w;
+}
+
 void S_Start(void)
 {
   int mnum;
@@ -208,7 +217,7 @@ void S_Start(void)
     mnum = idmusnum; //jff 3/17/98 reload IDMUS music if not -1
   else
     if (gamemode == commercial)
-      mnum = mus_runnin + gamemap - 1;
+      mnum = mus_runnin + WRAP(gamemap - 1, NUMMUSIC - mus_runnin);
     else
       {
         static const int spmus[] =     // Song - Who? - Where?
@@ -225,22 +234,14 @@ void S_Start(void)
         };

         if (gameepisode < 4)
-          mnum = mus_e1m1 + (gameepisode-1)*9 + gamemap-1;
+          mnum = mus_e1m1 + WRAP((gameepisode-1)*9 + gamemap-1, mus_runnin - mus_e1m1);
         else
-          mnum = spmus[gamemap-1];
+          mnum = spmus[WRAP(gamemap-1, 9)];
       }

   memset(&musinfo, 0, sizeof(musinfo));
 ```

Reply:
Doom 1 isn't as safe as I thought. Forgot the ExM0 and E0My cases.